### DISCUSSION TITLE
2 Layers blending and scheduling

### DISCUSSION THREAD

**Discussion by Doubl3MD** [halide/Halide#7308] [Q&A]
Hi fellow Haliders,

I'm trying to make a very simple layer blending pipeline.
It works like expected but when i try to add scheduling it is in fact slower than no scheduling at all.

The code is simple and should be self explanatory but to be sure:

I have 2 planar rgba input images(layers), bottom & top.

I loop over x y check if there is an alpha and use this in a branchless fashion to compute if the pixel of top is to be draw or otherwise the bottom pixel.

I was anticipating  a boost in performance with scheduling when testing this on 1000*1000(1 MP) pixel layers.  
But it is in fact slower, there probably is a logical explanation...?

Any hint on why and maybe how to get it to run faster would be more then welcome. :)

Thx in advance!

(Interleaving comes to mind for blending operations, but dono...)

```
ImageParam p_bottom(type_of<uint8_t>(), 3, "p_bottom");
 ImageParam p_top(type_of<uint8_t>(), 3, "p_top");

 Func blendLayers;
 Var x,y,c;

 Expr checkAlpha = cast<bool>(p_top(x, y, 3));

 blendLayers(x, y, c) = p_bottom(x, y, c) *!checkAlpha + checkAlpha * p_top(x, y, c);

//this "standard" scheduling is slower than no schedule at all
 Var x_outer, y_outer, x_inner, y_inner, tile_index;
 blendLayers
    .tile(x, y, x_outer, y_outer, x_inner, y_inner, 64, 64)
    .fuse(x_outer, y_outer, tile_index)
    .parallel(tile_index);

Var x_inner_outer, y_inner_outer, x_vectors, y_pairs;
 blendLayers
    .tile(x_inner, y_inner, x_inner_outer, y_inner_outer, x_vectors, y_pairs, 4, 2)
    .vectorize(x_vectors)
    .unroll(y_pairs);

```

**Comment by abadams** [comment]
A good default schedule for this sort of code is:
```
blendLayers.reorder(c, y, x).vectorize(x, 32).parallel(y, 8).bound(c, 0, 3 /* or maybe 4 */).unroll(c);
```
And I'd write blendLayers as:

```
blendLayers(x, y, c) = select(p_top(x, y, 3) != 0, p_top(x, y, c), p_bottom(x, y, c));
```
select is a branchless ternary operator

**Reply by Doubl3MD** [reply]
Thx for the reply. This schedule runs 4 times slower then no schedule... 

This makes me suspicious about if there is something wrong elsewhere... So I digged up some previously written Halide code that was certainly x8(core count) faster with scheduling. It turns out that that code als runs slower now when scheduled for vectorizing and parallel. 

I'm working with Visual Studio 2022. 
Just updated to latest update right before... dono if that could have caused the problems? 
Any change you know what could cause this kind of problem?

**Reply by Doubl3MD** [reply]
Ok after going back and forth... there is no problem with my setup and the previously written code runs a lot faster with scheduling. the Faulty test was caused by an oversight on my end.

So the proposed schedule is about 3 times slower then no schedule at all.
Is the parallelizing creating more overhead then the actual speedgain it should give?
Computation is extremely minimal in this case....

**Reply by abadams** [reply]
How big is your image?

**Reply by Doubl3MD** [reply]
I tested with 1024x1024 and to be sure also 4096x4096. 

In both cases any scheduling makes it worse.

Funny thing is that when I try tiles in parallel and do the operation hundreds of times the cpu goes to a lot more than one core in percentage: 70-80-ish %. I noticed that the kernel time is also way up. 

With no schedule its around 13% but it's faster in the end.

Also I did dig in the tutorials again. The "fast gradient" at the end of tutorial 5 also does not look to benefit much from scheduling.

I have a "11th Gen Intel(R) Core(TM) i7-11850H @ 2.50GHz" in my machine. Dono if it matters.
